eng/common/scripts/get-codeowners.ps1

```diff
@@ -1,49 +1,94 @@
+  [string]$ToolPath = (Join-Path ([System.IO.Path]::GetTempPath()) "codeowners-tool-path"), # The place to check the tool existence. Put temp path as default
+  [string]$DevOpsFeed = "https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json", # DevOp tool feeds.
+  [string]$VsoVariable = "", # Option of write code owners into devop variable
+  [switch]$Test  #Run test functions against the script logic
 )
-$target = $TargetDirectory.ToLower().Trim("/")
-$codeOwnersLocation = Join-Path $RootDirectory -ChildPath ".github/CODEOWNERS"
-$ownedFolders = @{}
 
-if (!(Test-Path $codeOwnersLocation)) {
-  Write-Host "Unable to find CODEOWNERS file in target directory $RootDirectory"
-  exit 1
-}
+$ToolCommandName = "retrieve-codeowners"
 
-$codeOwnersContent = Get-Content $codeOwnersLocation
+function Get-CodeOwnersTool()
+{
+  # Check if the retrieve-codeowners tool exsits or not.
+  if (Get-Command "$ToolPath/$ToolCommandName" -errorAction SilentlyContinue) {
+    return "$ToolPath/$ToolCommandName"
+  }
+  if (!(Test-Path $ToolPath)) {
+    New-Item -ItemType Directory -Path $ToolPath | Out-Null
+  }
+  Write-Warning "Installing the ToolCommandName tool under $ToolPath... "
+  dotnet tool install --tool-path $ToolPath --add-source $DevOpsFeed --version $ToolVersion "Azure.Sdk.Tools.RetrieveCodeOwners" | Out-Null
 
-foreach ($contentLine in $codeOwnersContent) {
-  if (-not $contentLine.StartsWith("#") -and $contentLine){
-    $splitLine = $contentLine -split "\s+"
-    
-    # CODEOWNERS file can also have labels present after the owner aliases
-    # gh aliases start with @ in codeowners. don't pass on to API calls
-    $ownedFolders[$splitLine[0].ToLower().Trim("/")] = ($splitLine[1..$($splitLine.Length)] `
-      | ? { $_.StartsWith("@") } `
-      | % { return $_.substring(1) }) -join ","
+  $command = Join-Path $ToolPath $ToolCommandName
+  # Test to see if the tool properly installed.
+  if (!(Get-Command $command -errorAction SilentlyContinue)) {
+    Write-Error "The retrieve-codeowners tool is not properly installed. Please check your tool path. $ToolPath"
+    return 
   }
+  return $command
 }
 
-$results = $ownedFolders[$target]
-
-if ($results) {
-  Write-Host "Found a folder $results to match $target"
+function Get-CodeOwners ([string] $command)
+{
+  if (!$command) {
+    return @()
+  }
+  # Params $RootDirectory is already in use in cpp release pipeline. 
+  # Will use $CodeOwnerFileLocation and deprecate $RootDirectory once it is ready to retire $RootDirectory.
+  if ($RootDirectory -and !(Test-Path $CodeOwnerFileLocation)) {
+    $CodeOwnerFileLocation = Join-Path $RootDirectory ".github/CODEOWNERS"
+  }
+  
+  $codeOwnersString = & $command --target-directory $targetDirectory --code-owner-file-path $CodeOwnerFileLocation
+  # Failed at the command of fetching code owners.
+  if ($LASTEXITCODE -ne 0) {
+    return @()
```

Comment:
Make sure you test this as I don't think it is doing what you expect. For powershell I think you might need to return `,@()`, see https://stackoverflow.com/questions/18476634/powershell-doesnt-return-an-empty-array-as-an-array